textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/queryrewriter/FuzzyTokenizer.java

```diff
@@ -0,0 +1,144 @@
+ * After recursive call returns with a set of parsed strings, prefix is appended in front of each string
+ */
+public class FuzzyTokenizer
+{
+    //Data members
+    private static WordBase wordBase;
+    private String phrase;
+
+    static {
+        try {
+            wordBase = new WordBase("wordsEn.txt");
+        }
+        catch(IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    /**
+     * Parameterized constructor requires phrase string that is to be tokenized
+     * Creates wordBase object containing knowledge base of words in file wordsEn.txt
+     * @param phrase
+     * @throws IOException
+     */
+    public FuzzyTokenizer(String phrase) {
+        this.phrase = phrase;
+    }
+
+    /**
+     * When called returns array of rewritten phrases
+     * Splits phrase string into space separated terms
+     * For each term performs fuzzy tokenization to create corresponding lists of rewritten terms
+     * Calls method rewriteTerm to perform tokenization over single term
+     * Performs a cross product concatenation of each rewritten term list and returns final String array
+     * Calls method crossCatenate to perform required concatenation of multiple lists
+     * @return
+     */
+    public List<String> getFuzzyTokens() {
+        String[] queryWords = phrase.split("\\s+");
+        List< List<String> > allWordsList = new ArrayList< List <String> >();
+        for(String word : queryWords)
+            allWordsList.add(rewriteTerm(word));
+        return crossCatenate(allWordsList);
+    }
+
+    /**
+     * Wrapper over main recursive method, rewrite which performs the fuzzy tokenization of a term
+     * Ensures that original phrase term is also included in returned rewritten terms list
+     * Is necessary if original phrase term is not a valid word according to word knowledge base
+     * @param term
+     * @return
+     */
+    private List<String> rewriteTerm(String term) {
+        List<String> queryWordList = rewrite(term);
+        if(! term.equals(queryWordList.get(queryWordList.size()-1)))
+            queryWordList.add(term);
+        return queryWordList;
+    }
+
+    /**
+     * Main recursive algorithm to find possible phrase strings that original phrase term can be mapped to
+     * Assumes original phrase term may have a tokenization mistake due to missing multiple space delimiters
+     * Tries to find all possible valid prefixes of original term as per word knowledge base
+     * For each prefix, sends rest of string to a recursive call to get all possible rewritten suffixes
+     * Attaches prefix term in front of each rewritten suffix in the list
+     * Returns final phrase term list
+     * @param term
+     * @return
+     */
+    private List<String> rewrite(String term) {
+        List<String> queryList = new ArrayList<String>();
+        for(int i=1; i<=term.length(); i++) {
+            if(i == term.length()) {
```

Comment:
Not exactly.  I wonder if we remove the "if" statement, does the remaining code cover this "if" case automatically?